### PR TITLE
fix(provider): Rewrite EVEOnline in TS, fix default scopes

### PR DESCRIPTION
**Pull request by mckernanin**
## Reasoning 💡

Rewrite the EVEOnline Provider in typescript, and provide the proper default scope.

## Checklist 🧢

- [x] Documentation
- [ ] Tests
- [ ] Ready to be merged

## Affected issues 🎟
[Manual testing of providers for v4](https://github.com/nextauthjs/next-auth/issues/2524)

### CODE REVIEW

File: src/providers/eveonline.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    id: "eveonline",
+    name: "EVE Online",
+    type: "oauth",
+    token: "https://login.eveonline.com/oauth/token",
+    userinfo: "https://login.eveonline.com/oauth/verify",
+    authorization: {
+      params: { scope: "publicData" },
+      url: "https://login.eveonline.com/oauth/authorize",
+    },
+    profile(profile: P) {
```
Comment (balazsorban44) [inline src/providers/eveonline.ts:27]:
```suggestion
    profile(profile) {
```

Type here is supposed be inferred.

File: src/providers/eveonline.ts [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      return {
+        id: profile.CharacterID,
+        name: profile.CharacterName,
+        email: null,
```
Comment (balazsorban44) [inline src/providers/eveonline.ts:31]: Any ideas if we could get the email also by default? maybe adding an `email` scope? Could you test it out? :pray:

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Just two tiny comments, but I appreciate you taking the time to fix this integration! :green_heart:

Do you know if there are any strange requirements to register an application on EVEOnline? I would like us to be able to test these integrations in the future.

File: src/providers/eveonline.ts (hunk 1 above)
Comment (mckernanin) [reply src/providers/eveonline.ts:27]: Sure. Should update the Google provider as well, that's where I pulled the base from.

**Comment by mckernanin** [comment]
I believe that you must have paid for something once in the account, so that they have a connection to a real person.

**Comment by balazsorban44** [comment]
That's less than ideal, but it means a good use of our OpenCollective money, maybe. 🙄
https://opencollective.com/nextauth

**Comment by mckernanin** [comment]
The cheapest thing available in their store is $2. A few things to note, that are peculiarities of EVE:

- Sharing accounts is not allowed, so the account created to use a set of credentials needs to be owned by an individual
- It's not possible to get emails from their oauth because oauth tokens are tied to characters on an account, not the actual user account. This is mainly to prevent doxxing.

Depending on what you want to have setup to test it, I would be willing to create a set of credentials, as long as they would not be shared publicly.

**Comment by mckernanin** [comment]
@balazsorban44 let me know what else I can do here to get this across the line :)

**Comment by balazsorban44** [comment]
Apologies, I have been busy with work and other stuff. :slightly_smiling_face: I would like to take on the offer of having an app we could test against, you can contact me by e-mail or on Twitter about the details! :green_heart: I greatly appreciate it!

We would only use that account to test the most basic login flow, and that the profile is filled out properly. I can talk about it in more detail when you contact me.

**Comment by mckernanin** [comment]
No worries just wanted to make sure you had seen my last. The app I'm getting next-auth setup in will be public that you can use for testing. I tried to DM you on Twitter but it doesn't seem they're open -- feel free to toss me a dm [@kevinmckernan](https://twitter.com/kevinmckernan)

**Comment by mckernanin** [comment]
Well, fun update on this. They've just announced some breaking changes for the EVE SSO API: https://developers.eveonline.com/blog/article/sso-endpoint-deprecations-2

I'm working on writing a no-framework node implementation for their example repo. The TLDR is:

- Access tokens are now JWTs
- Verify endpoint is deprecated
- JWTs are verifiable using a JWKS endpoint

I'm not sure if there's an existing flow like this in next-auth somewhere, can you point me in the right direction?

**Comment by mckernanin** [comment]
@balazsorban44 poke on above question when you have time, thanks!

**Comment by mckernanin** [comment]
Ugh sorry about that on the blog link, yeah it's broken apparently 🙄 
Here's a wayback machine article:

https://web.archive.org/web/20211009142752/https://developers.eveonline.com/blog/article/sso-endpoint-deprecations-2

**Comment by balazsorban44** [comment]
it's an error at the provider level. access_token should not be used as id_token, but it's easy to override the Adapter method that creates the account.

**Comment by mckernanin** [comment]
Agreed that it's a provider issue. I'll look for that Adapter override.

**Comment by balazsorban44** [comment]
@mckernanin so is there anything remaining here that you think blocks us from merging this?

A proof of login with name, email, and a picture would be nice (or whatever that is returned of those)

**Comment by lluia** [comment]
I was trying to test this; however, it seems you need to pay for a developer account in **EVE online** 🥲

Screenshot 2021-11-18 at 09 04 26

I'll wait for what @mckernanin says ☕️

**Comment by mckernanin** [comment]
Hey all, I've been super busy at work with little free time for code. We've got a holiday this week in the US and I'll get back to this. I want to figure out how to recommend setting up the adapter as Balazs and I discussed. I need to finish my application using it, which I'm pretty close on.

**Review by balazsorban44 — APPROVED** [review]

**Pull request merged**